File: 6 Zigzag Conversion/C/solution.c
#include <stdlib.h>
#include <string.h>

char* convert(char* s, int numRows) {
    size_t len = strlen(s);

    if (numRows == 1 || len <= numRows) {
        return s;
    }

    char* result = (char*) malloc(sizeof(char) * (len + 1));
    int resultLen = 0;

    int max = 2 * (numRows - 1);
    for (int i = 0; i < numRows; i++) {
        for (int j = i; j < len; j += max) {
            result[resultLen] = s[j];
            resultLen++;

            if (i != 0 && i != numRows - 1) {
                int secondIdx = j + max - 2 * i;
                if (secondIdx < len) {
                    result[resultLen] = s[secondIdx];
                    resultLen++;
                }
            }
        }
    }

    result[resultLen] = '\0';
    return result;
}

File: 6 Zigzag Conversion/Python/Solution.py
from typing import List


class Solution:
    def convert(self, s: str, numRows: int) -> str:
        l: int = len(s)
        
        if numRows == 1 or l <= numRows:
            return s
        
        result: str = ""
        max: int = 2 * (numRows - 1)
        for i in range(numRows):
            j: int = i
            while j < l:
                result += s[j]
                
                if i != 0 and i != numRows - 1:
                    secondIdx: int = j + max - 2 * i
                    if secondIdx < l:
                        result += s[secondIdx]
                
                j += max
        
        return result